### PR TITLE
Phantom let support in Cmm

### PR DESCRIPTION
**Pull request by mshinwell**
This patch adds the "phantom let" construct to the Cmm language and propagates it through `Cmmgen`.

@lthls Perhaps you could review this, it follows on from GPR#2060.

### CODE REVIEW

**Review by lthls — APPROVED** [review]
A straight-forward propagation of the phantom-let constructions to Cmm. These constructions are dropped at Selectgen for now, but I assume a subsequent PR will build on this one and propagate them further.

I'm wondering about reusing the exact same type as Clambda for defining expressions; this is certainly convenient, but it might be cleaner in this case to define a specific type for the constants allowed to appear in `Uphantom_const` and `Uphantom_read_symbol_field`.
Do you think it would simplify the code that later deconstructs these expressions ?
It would also allow you to put these type definitions into their own file, and simplify dependencies a bit.

**Comment by mshinwell** [comment]
I'm happy to use a separate type; indeed I was wondering about that myself.  I think in general it's probably a mistake to share types across intermediate languages, even though it does require rather more typing to establish a situation where they are not shared.

**Comment by mshinwell** [comment]
@lthls Please look at `8913f42`.  If that looks reasonable then I will adapt GPR#2071 to make use of this.  I don't think we need to use a separate file for that time at this juncture.

**Comment by lthls** [comment]
This looks better indeed.
I still have two questions:
- Is there a reason why the Clambda version of `Uphantom_read_symbol_field` takes its symbol in the form of a `uconstant` instead of a bare string ? I assumed the associated defining expr could be useful later, but if you don't use it when translating to Cmm, you might as well forbid it as early as Clambda.
- Could you document that the constant integers used in the Cmm `Cphantom_const_int` represent OCaml integers, untagged ? This is obvious in earlier representations, but Cmm mostly manipulates machine integers, so I think a comment on the meaning/encoding here wouldn't hurt (of course, a full conversion of the backend to use the Targetint types would help, too, but that's outside the scope of this PR).

**Comment by mshinwell** [comment]
I don't remember why `Uphantom_read_symbol_field` didn't just take a symbol name, so I've changed it to that for the moment.  We can always alter it later.

Good point about the tagging---in fact, `Cphantom_const_int` should just be like `Cconst_int`.  Previously, the DWARF emitting code was doing the tagging, which is wrong.  Now it's done in `Cmmgen`.  I've also changed this constructor to use `Targetint.t`.  (It should really be "target-width integers of the same size as [int]", like `Targetint.OCaml` in the new Flambda branch, but we don't have that type merged yet.)

**Comment by chambart** [comment]
I trust @lthls review on this, and approve. This is good to merge as soon as the conflict on Changes is fixed.

**Comment by mshinwell** [comment]
Going to wait for CI on this one before merging.

**Pull request merged**
